### ISSUE TITLE
[Feature] [Windows] - Behavior of tailscale regarding startup / autostart

### ISSUE THREAD

**Issue by tryallthethings** [tailscale/tailscale#983]
Could you please add an option to the Windows Tailscale installer to ask the user whether he / she wants the Tailscale GUI to start on system boot or not? I couldn't find the NSIS configuration in the repo to directly make a pull request, so I assume that's part of the closed source of the GUI clients.
The option can be default on the keep the current behavior.

Thank you!

**Comment by alexbrainman** [comment]
> whether he / she wants the Tailscale GUI to start on system boot or not?

@tryallthethings

Tailscale GUI starts when user login to the desktop, not when system boots. So, perhaps, you mean something different. Can you explain it for me, please?

Thank you.

Alex

**Comment by tryallthethings** [comment]
Yes, that's what I meant. Sorry, I'm mostly working on a single user PC currently so for me it's the same. :see_no_evil:
The setup adds a shortcut to Tailscale to the Startup folder:

![image](https://user-images.githubusercontent.com/11146296/100926627-0a01a580-34e4-11eb-8d6a-9960b7f1fcad.png)
 `Create shortcut: C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Startup\Tailscale.lnk`

I'd like this step to be optional during the setup process, so I don't have to remove it after every update.

**Comment by alexbrainman** [comment]
> The setup adds a shortcut to Tailscale to the Startup folder:
> 
> ...
> `Create shortcut: C:\ProgramData\Microsoft\Windows\Start Menu\Programs\Startup\Tailscale.lnk`
> 
> I'd like this step to be optional during the setup process, so I don't have to remove it after every update.

Thanks for explaining.

If you don't have Tailscale GUI running, how would you configure the Tailscale service?

Also, how difficult it is to remove Tailscale GUI from Startup folder? Adding new option to the installer that everyone will see and have to click on `Next` button is the downside of your suggestion. What is the upside?

Thank you.

Alex

**Comment by tryallthethings** [comment]
I'm only using Tailscale occasionally on my Windows machine and start the GUI once I need it manually.

It's not difficult to remove. But it's inconvenient that I have to re-do it with each Tailscale update.

IMHO it's not good practice to force additional programs into the user's autostart (same goes for placing program shortcuts on the user's desktop) without giving the user a choice. I've heard way to many user complaints about slow computers with a hundred unnecessary programs / services running and filling up the system tray.
If you don't want another "Next" screen you could add a check during installation if the startup shortcut was previously removed and not add it back.

**Comment by bradfitz** [comment]
I agree that the installer should ask that question.

We have some other installer work coming up, so we'll keep this in mind for the same time.

**Comment by alexbrainman** [comment]
> I'm only using Tailscale occasionally on my Windows machine and start the GUI once I need it manually.
> 
> It's not difficult to remove. But it's inconvenient that I have to re-do it with each Tailscale update.
> 
> IMHO it's not good practice to force additional programs into the user's autostart (same goes for placing program shortcuts on the user's desktop) without giving the user a choice. I've heard way to many user complaints about slow computers with a hundred unnecessary programs / services running and filling up the system tray.

Fair enough. Thanks for explaining.

> If you don't want another "Next" screen you could add a check during installation if the startup shortcut was previously removed and not add it back.

I am not sure this would work very well if you removed the shortcut. And installer does not restore the shortcut back. Too magical for me.

Alex

**Comment by tryallthethings** [comment]
> I agree that the installer should ask that question.
> 
> We have some other installer work coming up, so we'll keep this in mind for the same time.
Sounds good!

> > If you don't want another "Next" screen you could add a check during installation if the startup shortcut was previously removed and not add it back.
> 
> I am not sure this would work very well if you removed the shortcut. And installer does not restore the shortcut back. Too magical for me.
Sorry, I don't get that?

**Comment by alexbrainman** [comment]
> Sorry, I don't get that?

What you propose (as far as understand your proposal), is to have Tailscale installer automatically determine if Tailscale GUI is configured to automatically run on user logon. And, if not configured, not to configure GUI to run after install finishes.

Let's say that user needs to run GUI for some reason. He finds it on the disk and runs the GUI.

Next time he needs to run the GUI, he is slightly annoyed that GUI is not running, but he finds it on the disk again and runs it.

Next time he needs the GUI he is really annoyed that it is not running. So he runs the installer in the hope that installer will help him. But the installer won't help him in this situation.

So the user needs to know / remember to configure Tailscale GUI to auto-start. Otherwise they are stuck.

I hope it is clear enough.

Alex

**Comment by tryallthethings** [comment]
I suppose if a user is able to remove the shortcut from the startup folder he could add it back as well.

But I could propose a more sophisticated solution as well. 

1. Remove adding of the autostart shortcut from the installer. 
2. Ask the user on first startup of the GUI if he wants it to be started automatically and add the startup shortcut if necessary
3. Add a configuration item "start automatically on system boot" to the context menu:
![image](https://user-images.githubusercontent.com/11146296/101177675-6087f480-3648-11eb-8e4e-168bb28b8502.png)
4. Save the user selection to prefs.conf, so it is persistent throughout updates.

**Comment by alexbrainman** [comment]
> 3\. Add a configuration item "start automatically on system boot" to the context menu:

What do you propose Tailscale GUI do when this context menu gets ticked? Please describe both "tick-on" and "tick-off" scenarios.

> 4\. Save the user selection to prefs.conf, so it is persistent throughout updates.

Which program will read this prefs.conf setting? And how that program behaviour will change depending on this setting?

Thank you.

Alex

**Comment by tryallthethings** [comment]
> > 3. Add a configuration item "start automatically on system boot" to the context menu:
> 
> What do you propose Tailscale GUI do when this context menu gets ticked? Please describe both "tick-on" and "tick-off" scenarios.

Tick on: Create shortcut to Tailscale GUI in startup folder. Set a parameter like "Autostart" to true in the prefs.conf.
Tick off: Remove the shortcut. Set a parameter like "Autostart" to false in the prefs.conf.

> > 4. Save the user selection to prefs.conf, so it is persistent throughout updates.
> 
> Which program will read this prefs.conf setting? And how that program behaviour will change depending on this setting?
> 

What do you mean with "Which program will read this prefs.conf setting?" What other program besides Tailscale should read it?prefs.conf is the Tailscale configuration file in %LOCALAPPDATA%\Tailscale where settings are saved.

**Comment by alexbrainman** [comment]
> What do you mean with "Which program will read this prefs.conf setting?" What other program besides Tailscale should read it?prefs.conf is the Tailscale configuration file in %LOCALAPPDATA%\Tailscale where settings are saved.

If Tailscale GUI removes itself from startup folder (as you suggest in the previous message), it won't be running next time user login. And next time. And next. It is irrelevant what prefs.conf contains. How do you propose user handles this situation?

There is also problem where proposed "Autostart" flag setting in prefs.conf gets out of sync with "is Tailscale GUI in startup folder or not". Perhaps it is not important. But still feels confusing to me.

Anyway. Thanks for your input. I will let @bradfitz decide what to do here.

Alex

**Comment by tryallthethings** [comment]
> If Tailscale GUI removes itself from startup folder (as you suggest in the previous message), it won't be running next time user login. And next time. And next. It is irrelevant what prefs.conf contains. How do you propose user handles this situation?

Lauch Tailscale from the start menu. And if the user wants it to start automatically: click the respective button in the context menu

**Comment by DentonGentry** [comment]
Since this issue was filed, a "tailscale logout" command was added. The mechanism we advise now is to use "tailscale logout" as a logoff script. That way the next time you login to Windows, Tailscale will not automatically connect.

This is an example of installing a logoff script:
https://technology.siprep.org/create-an-auto-logoff-script-for-windows-7/

----

Barring further updates, we'll expect to close this during the next bug scrub.

**Comment by tryallthethings** [comment]
> Since this issue was filed, a "tailscale logout" command was added. The mechanism we advise now is to use "tailscale logout" as a logoff script. That way the next time you login to Windows, Tailscale will not automatically connect.
> 
> This is an example of installing a logoff script:
> https://technology.siprep.org/create-an-auto-logoff-script-for-windows-7/
> 
> ----
> 
> Barring further updates, we'll expect to close this during the next bug scrub.

Your comment is unfortunately completely unrelated to this issue. This is about installer behavior.

**Comment by DentonGentry** [comment]
I'm not convinced that adding more checkboxes in the installer is the right thing to do. That is the very first glimpse of the Tailscale product, having more complexity isn't clear to me as being a good thing to do.

If one doesn't want Tailscale to connect at boot, "tailscale logout" will prevent it from doing so.

**Comment by DiegoFleitas** [comment]
I agree with tryallthethings, setting up new software to automatically start during start up and not easily allowing the user to turn it off is a dark pattern and thats kind of a red flag for me (specially since its software you might install on multiple machines you own / colleagues machines). 
Many programs that start on windows start up, allow you to change that behaviour through settings.
If this a technical issue I think it could be solved by adding the suggested checkbox, checked by default.

**Comment by lmcnearney** [comment]
👍 Any progress on this issue? As a new Tailscale user who only needs to run it periodically, I found the autostart behavior annoying and there is no permanent way to toggle it off given that the installer always restores the shortcut in the startup folder. Most programs would provide a checkbox during the installation and also allow a config option (likely through the right-click context menu [as suggested](https://github.com/tailscale/tailscale/issues/983#issuecomment-738828734).

**Comment by bradfitz** [comment]
cc @dblohm7

**Comment by DentonGentry** [comment]
@lmcnearney does "Log out" fit your use?
<img width="315" alt="Screen Shot 2021-11-22 at 10 26 46 AM" src="https://user-images.githubusercontent.com/1140756/142915443-38772e29-1401-425e-80b0-3a765cedb65d.png">

**Comment by lmcnearney** [comment]
> @lmcnearney does "Log out" fit your use?

Only if there was an option to automatically log out when the GUI is closed. This would make it behave more like classic VPN clients which prompt for credentials on every run. An option to automatically connect or disconnect would be my personal preference to avoid having Tailscale pop up a notification on startup and having to login every time.

These options won't resolve the underlying issue of having the GUI automatically run on startup however. As others have commented, it's more about respect for allowing a user to control what runs on their system. There is no choice given and the installer's behavior of automatically added itself on every update even overrides my manual intervention of deleting that shortcut.

**Comment by Itamal** [comment]
Hello! I've found this thread helpful. It's not all that hard to disable startup behavior from the windows menu, even though it's not common behavior on windows apps. Perhaps, a link from the right-click menu to `ms-settings:startupapps ` can direct unaware users to the setting location.

**Comment by hutch85** [comment]
Echoing others here and the sentiment towards automatically connecting to Tailscale. It's one thing to autostart the GUI client and that's fine, but it's a separate situation over being able to specify *automatically connecting* at Windows startup.

Disabling the client at startup or running scripts is rather silly at best - many other applications provide the means and setting to have an option to manage connection preferences upon launch.

**Comment by rasupit** [comment]
I would also like to see an option to automatically connect to tailscale as an option.

![image](https://github.com/user-attachments/assets/5878ceac-9884-4658-b63e-e9443b1a12bb)

**Comment by arty-hlr** [comment]
Five years after this issue was opened, I was also surprised that tailscale is auto-start by default and there is no easy way to turn that off, outside of logging out I guess.

**Comment by Dan-Piz** [comment]
Echoing the other comments, there needs to be a way - during install and updates - to decide whether the user wants Tailscale to start automatically. This should be for the service entirely and maybe the GUI separately. In our use case we do not want Tailscale running whatsoever unless the user decides to run it.
